File: app/src/main/java/com/example/Message.kt
package com.example.model

enum class MessageType {
    TEXT,
    IMAGE,
    VIDEO,
    VOICE,
    FILE,
    LOCATION,
    LIVE_LOCATION,
    VIEW_ONCE_IMAGE,
    VIEW_ONCE_VIDEO,
    SYSTEM_KEY_CHANGE
}

/**
 * Persistence/network-facing status values kept for source compatibility with the existing app.
 * The current server only confirms acceptance, so SENT must never be rendered as delivered/read.
 */
enum class MessageStatus {
    QUEUED,
    SENDING,
    SENT,
    DELIVERED,
    READ,
    FAILED
}

/**
 * Explicit product semantics for the UI. DELIVERED/READ are intentionally future-facing: the
 * current backend does not emit those receipts, therefore production mapping stops at
 * SENT_TO_SERVER.
 */
enum class MessageDeliveryState {
    QUEUED_OFFLINE,
    SENDING,
    SENT_TO_SERVER,
    DELIVERED_TO_RECIPIENT_DEVICE,
    READ,
    FAILED
}

fun MessageStatus.toDeliveryState(): MessageDeliveryState = when (this) {
    MessageStatus.QUEUED -> MessageDeliveryState.QUEUED_OFFLINE
    MessageStatus.SENDING -> MessageDeliveryState.SENDING
    MessageStatus.SENT -> MessageDeliveryState.SENT_TO_SERVER
    MessageStatus.DELIVERED -> MessageDeliveryState.DELIVERED_TO_RECIPIENT_DEVICE
    MessageStatus.READ -> MessageDeliveryState.READ
    MessageStatus.FAILED -> MessageDeliveryState.FAILED
}

data class Message(
    val id: String,
    val conversationId: String,
    val senderId: String,
    val senderName: String,
    val text: String,
    val type: MessageType = MessageType.TEXT,
    // Safe default for a model materialized from the server. Local optimistic sends set SENDING.
    val status: MessageStatus = MessageStatus.SENT,
    val timestamp: String,
    val timestampMillis: Long = System.currentTimeMillis(),
    val mediaUrl: String? = null,
    val mediaSize: String? = null,
    val voiceDurationSeconds: Int = 0,
    val voiceWaveform: List<Float> = emptyList(),
    val isViewOnceOpened: Boolean = false,
    val replyToText: String? = null,
    val reactions: List<String> = emptyList(),
    val latitude: Double? = null,
    val longitude: Double? = null,
    val fileName: String? = null,
    val fileSize: String? = null,
    val fileExtension: String? = null,
    val attachmentId: String? = null,
    val mimeType: String? = null,
    val attachmentKeyBase64: String? = null,
    val attachmentNonceBase64: String? = null,
    val localUri: String? = null,
    val isPinned: Boolean = false
)

File: flutter/lib/features/messages/domain/message.dart
enum ChatNuMessageType {
  text,
  image,
  video,
  voice,
  file,
  location,
  liveLocation,
  viewOnceImage,
  viewOnceVideo,
  system,
}

enum MessageDeliveryState {
  queuedOffline,
  sending,
  sentToServer,
  deliveredToRecipientDevice,
  read,
  failed,
}

class ChatNuMessage {
  const ChatNuMessage({
    required this.id,
    required this.conversationId,
    required this.senderId,
    required this.senderName,
    required this.body,
    required this.sentAt,
    this.clientId,
    this.type = ChatNuMessageType.text,
    this.deliveryState = MessageDeliveryState.sentToServer,
    this.attachmentId,
    this.fileName,
    this.mimeType,
    this.sizeBytes,
    this.attachmentKeyBase64,
    this.attachmentNonceBase64,
    this.locationLatitude,
    this.locationLongitude,
    this.mediaDurationMs,
    this.isVideoNote = false,
  });

  final String id;
  final String? clientId;
  final String conversationId;
  final String senderId;
  final String senderName;
  final String body;
  final DateTime sentAt;
  final ChatNuMessageType type;
  final MessageDeliveryState deliveryState;
  final String? attachmentId;
  final String? fileName;
  final String? mimeType;
  final int? sizeBytes;
  final String? attachmentKeyBase64;
  final String? attachmentNonceBase64;
  final double? locationLatitude;
  final double? locationLongitude;
  final int? mediaDurationMs;
  final bool isVideoNote;

  bool get hasAttachment => attachmentId != null;
  bool get hasLocation => locationLatitude != null && locationLongitude != null;
  bool get isPlayableAudio =>
      type == ChatNuMessageType.voice ||
      (mimeType?.toLowerCase().startsWith('audio/') ?? false);

  ChatNuMessage copyWith({
    String? id,
    String? clientId,
    String? body,
    DateTime? sentAt,
    ChatNuMessageType? type,
    MessageDeliveryState? deliveryState,
    String? attachmentId,
    String? fileName,
    String? mimeType,
    int? sizeBytes,
    String? attachmentKeyBase64,
    String? attachmentNonceBase64,
    double? locationLatitude,
    double? locationLongitude,
    int? mediaDurationMs,
    bool? isVideoNote,
  }) {
    return ChatNuMessage(
      id: id ?? this.id,
      clientId: clientId ?? this.clientId,
      conversationId: conversationId,
      senderId: senderId,
      senderName: senderName,
      body: body ?? this.body,
      sentAt: sentAt ?? this.sentAt,
      type: type ?? this.type,
      deliveryState: deliveryState ?? this.deliveryState,
      attachmentId: attachmentId ?? this.attachmentId,
      fileName: fileName ?? this.fileName,
      mimeType: mimeType ?? this.mimeType,
      sizeBytes: sizeBytes ?? this.sizeBytes,
      attachmentKeyBase64: attachmentKeyBase64 ?? this.attachmentKeyBase64,
      attachmentNonceBase64:
          attachmentNonceBase64 ?? this.attachmentNonceBase64,
      locationLatitude: locationLatitude ?? this.locationLatitude,
      locationLongitude: locationLongitude ?? this.locationLongitude,
      mediaDurationMs: mediaDurationMs ?? this.mediaDurationMs,
      isVideoNote: isVideoNote ?? this.isVideoNote,
    );
  }
}
